### ISSUE TITLE
Improving Argument.of() for generics

### ISSUE THREAD

**Issue by lightoze** [micronaut-projects/micronaut-core#749]
Current signature `<T> Argument<T> of(Class<T> type, @Nullable Class<?>... typeParameters)` provides no way to obtain types like `Argument<List<String>>` without casting and getting compiler warnings. This is quite a bummer when working with HttpClient.

What do you think about changing it to `<T> Argument<T> of(Class<? super T> type, @Nullable Class<?>... typeParameters)`?

**Comment by graemerocher** [comment]
sure, seems reasonable

**Comment by lightoze** [comment]
Probably this should be done for all `Argument.of` variants involving type parameters, then, for consistency.

**Comment by graemerocher** [comment]
I think it may be better to have `Argument.ofList(Book.class)` to fully resolve this

**Comment by lightoze** [comment]
Having that would probably resolve 80% of the cases and help with typeless variable declarations (Java 11 var, Kotlin). However with only `ofList` it will be still impossible to obtain `Argument<A<T>>` wihout casting.

**Comment by graemerocher** [comment]
yeah I don't think you can achieve `Argument<A<T>>` with the current `of` signature. It would need something like https://docs.oracle.com/javaee/6/api/javax/ws/rs/core/GenericEntity.html in JAX-RS

**Comment by lightoze** [comment]
There's also a similar `TypeReference` in Jackson and `TypeToken` in Guava. `TypeToken` combines both API styles `TypeToken.of()` and `new TypeToken<>()`. Micronaut's `Argument` is an interface, but how about creating `abstract class GenericArgument<T> implements Argument<T>` with similar semantics?

**Comment by graemerocher** [comment]
That seems like a good idea
